File: backend/app/models/attendance_procedure.py
import enum
from sqlalchemy import Column, Integer, String, Text, ForeignKey, DateTime, BigInteger, Table, Enum
from sqlalchemy.orm import relationship
from app.core.database import Base
from app.models.mixins import TimestampMixin

# Junction table for AttendanceProcedure and Medication
attendance_procedure_medications = Table(
    "attendance_procedure_medications",
    Base.metadata,
    Column("attendance_procedure_id", BigInteger, ForeignKey("attendance_procedures.id"), primary_key=True),
    Column("medication_id", Integer, ForeignKey("medications.id"), primary_key=True),
)

class AttendanceProcedureStatus(str, enum.Enum):
    pending = "pending"
    in_progress = "in_progress"
    done = "done"
    canceled = "canceled"

class AttendanceProcedure(Base, TimestampMixin):
    __tablename__ = "attendance_procedures"
    
    id = Column(BigInteger, primary_key=True, index=True)
    attendance_id = Column(BigInteger, ForeignKey("attendances.id"), nullable=False)
    procedure_id = Column(Integer, ForeignKey("procedures.id"), nullable=False)
    
    status = Column(
        Enum(AttendanceProcedureStatus, values_callable=lambda obj: [e.value for e in obj], native_enum=False), 
        nullable=False, 
        server_default="pending"
    )
    description = Column(Text, nullable=True)
    
    start_time = Column(DateTime, nullable=True)
    end_time = Column(DateTime, nullable=True)
    
    ordered_by_id = Column(BigInteger, ForeignKey("employees.id"), nullable=True)
    executed_by_id = Column(BigInteger, ForeignKey("employees.id"), nullable=True)
    
    attendance = relationship("Attendance", back_populates="procedures")
    procedure = relationship("Procedure")
    
    ordered_by = relationship("Employee", foreign_keys=[ordered_by_id])
    executed_by = relationship("Employee", foreign_keys=[executed_by_id])
    
    medications = relationship("Medication", secondary=attendance_procedure_medications)

File: frontend/src/types/attendance_procedure.ts
import { Procedure } from "./procedure";
import { Employee } from "./employee";
import { Medication } from "./medication";

export enum AttendanceProcedureStatus {
  pending = "pending",
  in_progress = "in_progress",
  done = "done",
  canceled = "canceled",
}

export interface AttendanceProcedure {
  id: number;
  attendance_id: number;
  procedure_id: number;
  procedure: Procedure;
  status: AttendanceProcedureStatus;
  description?: string | null;
  start_time?: string;
  end_time?: string;
  ordered_by_id?: number;
  ordered_by?: Employee;
  executed_by_id?: number;
  executed_by?: Employee;
  medications: Medication[];
  created_at: string;
  updated_at: string;
}
